pkg/applicationserver/io/pubsub/redis/registry.go

```diff
@@ -0,0 +1,262 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/gogo/protobuf/proto"
+	pubsubunique "go.thethings.network/lorawan-stack/pkg/applicationserver/io/pubsub/unique"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	ttnredis "go.thethings.network/lorawan-stack/pkg/redis"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+)
+
+var (
+	errInvalidFieldmask   = errors.DefineInvalidArgument("invalid_fieldmask", "invalid fieldmask")
+	errInvalidIdentifiers = errors.DefineInvalidArgument("invalid_identifiers", "invalid identifiers")
+)
+
+// appendImplicitPubSubGetPaths appends implicit ttnpb.ApplicationPubSub get paths to paths.
+func appendImplicitPubSubGetPaths(paths ...string) []string {
+	return append(append(make([]string, 0, 3+len(paths)),
+		"created_at",
+		"ids",
+		"updated_at",
+	), paths...)
+}
+
+func applyPubSubFieldMask(dst, src *ttnpb.ApplicationPubSub, paths ...string) (*ttnpb.ApplicationPubSub, error) {
+	if dst == nil {
+		dst = &ttnpb.ApplicationPubSub{}
+	}
+	return dst, dst.SetFields(src, paths...)
+}
+
+// PubSubRegistry is a Redis PubSub registry.
+type PubSubRegistry struct {
+	Redis *ttnredis.Client
+}
+
+func (r *PubSubRegistry) allKey(ctx context.Context) string {
+	return r.Redis.Key("all")
+}
+
+func (r *PubSubRegistry) appKey(uid string) string {
+	return r.Redis.Key("uid", uid)
+}
+
+func (r *PubSubRegistry) idKey(appUID, id string) string {
+	return r.Redis.Key("uid", appUID, id)
+}
+
+func (r *PubSubRegistry) makeIDKeyFunc(appUID string) func(id string) string {
+	return func(id string) string {
+		return r.idKey(appUID, id)
+	}
+}
+
+// Get implements pubsub.Registry.
+func (r PubSubRegistry) Get(ctx context.Context, ids ttnpb.ApplicationPubSubIdentifiers, paths []string) (*ttnpb.ApplicationPubSub, error) {
+	pb := &ttnpb.ApplicationPubSub{}
+	if err := ttnredis.GetProto(r.Redis, r.idKey(unique.ID(ctx, ids.ApplicationIdentifiers), ids.PubSubID)).ScanProto(pb); err != nil {
+		return nil, err
+	}
+	return applyPubSubFieldMask(nil, pb, appendImplicitPubSubGetPaths(paths...)...)
+}
+
+var errApplicationUID = errors.DefineCorruption("application_uid", "invalid application UID `{application_uid}`")
+
+// Range implements pubsub.Registry.
+func (r PubSubRegistry) Range(ctx context.Context, paths []string, f func(context.Context, ttnpb.ApplicationIdentifiers, *ttnpb.ApplicationPubSub) bool) error {
+	uids, err := r.Redis.SMembers(r.allKey(ctx)).Result()
+	if err != nil {
+		return err
+	}
+	for _, uid := range uids {
+		appUID, psID := pubsubunique.ToUIDs(uid)
```

Comment:
This should be `ToPubsubID()` right?

Reply:
This returns both the PubSubID and the Application Unique ID.

Reply:
See below